### ISSUE TITLE
Add `findOrNull` method

### ISSUE THREAD

**Issue by skyflyer55** [prisma/photonjs#89]
Using the demo project 'graphql-apollo-server' typescript project, I query on post and use a dummy id argument and expect to get a null post obj back, but instead I get an internal server error.

```
t.field('post', {
      type: 'Post',
      nullable: true,
      args: { id: idArg() },
      resolve: (parent, { id }, ctx) => {
        return ctx.photon.posts.findOne({
          where: {
            id,
          },
        })
      },
    })
```
 ```
Query:

{
  post(id: "123") {
    id
    author {
      id
    }
  }
}

Result: Error in Photon: \n[\n  {\n \"error\": \"ConnectorError(RecordDoesNotExist)\"\n  }\n]"
```

Is there a way to tell photon to just send back **null** or is this a **bug**?

**Comment by schickling** [comment]
Thanks a lot for opening this issues @skyflyer55. While this behavior is intentional (i.e. not a bug), we've internally considered whether we should additionally introduce a `photon.posts.findOrNull` method. Would this solve your issue?

**Comment by skyflyer55** [comment]
@schickling That method would solve my issue.

**Comment by schickling** [comment]
Great to know! We'll add this feature in one of the upcoming releases. In the meanwhile you can use a try/catch as a workaround. :)

**Comment by schickling** [comment]
This is currently being specced out here: https://github.com/prisma/specs/pull/5
Once the spec is ready, this feature will be ready for implementation. 👌

**Comment by reinvanimschoot** [comment]
Can you elaborate on why this implementation was chosen? I'm having difficulties seeing the reasoning behind no record being found resulting an error. Why not just make this function return null when no record was found?

**Comment by auderer** [comment]
The default behavior should be to return `null`, not throw an error. Here's my thoughts on why:

1. The most popular ORMs, Sequelize and Mongoose, return `null` without throwing an error for `findOne`. See examples for [mongoose](https://mongoosejs.com/docs/api.html#model_Model.findOne) and [sequelize](http://docs.sequelizejs.com/manual/models-usage.html).

2. Throwing an error by default encourages [coding by exception](https://en.wikipedia.org/wiki/Coding_by_exception). This adds complexity to an otherwise simple, common use case. 

```
try {
  const user = await ctx.photon.findOne({ where: { email } })
  if (!user) {
    throw new Error('No user with that email!')
  }
} catch (err) {
  // handle server errors
}
```

```
let user
try {
  user = await ctx.photon.findOne({ where: { email } })
} catch (err) {
  if (err instanceof RecordDoesNotExistError) {
    throw new Error('No user with that email!')
  } else {
    // handle server errors
  }
}
```

Adding `findOrNull` will allow using the first example, but my argument is that throwing an error is a bad default. Instead, I would propose using `findOrThrow` which would throw an error if not found.

**Comment by auderer** [comment]
The default behavior should be to return `null`, not throw an error.

The most popular ORMs, Sequelize and Mongoose, return `null` without throwing an error for `findOne`. See examples for [mongoose](https://mongoosejs.com/docs/api.html#model_Model.findOne) and [sequelize](http://docs.sequelizejs.com/manual/models-usage.html). 

Instead, I propose using findOrThrow which would throw an error if not found. This would be more consistent with other ORMs and result in simpler code in most cases.

**Comment by matthewmueller** [comment]
Thanks for bringing this up! 

Personally, I still think `findOne` by default should throw when there's no records. I find that in practice when you don't find what you're looking for, it's an error and you can't proceed. There are many exceptions to this which is why we'll also need a `findOrNull`. If it returns `null`, this means that the caller always has to check `if (!record) { ... }` and if they forget (working in JS or Go), it's a runtime error.

To help with the discussion, do you mind looking through your codebases at the other ORMs and count how many times the `if (record === null)` conditional was useful?  We can use this data to decide what the default should be. 🙏

**Comment by victordombrovskiy** [comment]
In my previous project on Kotlin I had the following set of methods:

**Comment by gihrig** [comment]
In the prisma1/prisma-binding version of my current project a search for `db.exists` reveals 21 results in 6 files.

These entries generally take the form:

```javascript
  async updateCollection(_, args, ctx, info) {
...
    const collectionExists = await ctx.db.exists.Collection({
...
    if (!collectionExists) {
      throw new Error(`Collection not found`)
    }
...
```

I agree with @auderer that findOne should behave similarly to other common ORM libraries. It's the generally expected behavior. Photon _is_ unique, but that's no reason why it should be _different_.

For me it comes down to "is `record not found` expected or exceptional?" I would argue that, particularly where user input is used as a search criteria, record not found is expected, not exceptional.

Correspondingly, Photon should not throw an exception in this case.

As a Photon consumer, it makes the difference between cleaner code and using try/catch. Try/catch tends to obscure what the code will do, making more work for the developer.

Example code from [photonjs typescript graphql-auth example](https://github.com/prisma/photonjs/blob/master/examples/typescript/graphql-auth/src/resolvers/Mutation.ts)
<details>
<summary></summary><p>
Example - cleaner:

```typescript
    t.field('login', {
      type: 'AuthPayload',
      args: {
        email: stringArg(),
        password: stringArg(),
      },
      resolve: async (parent, { email, password }, context) => {
        const user = await context.photon.users.findOne({
          where: {
            email,
          },
        })
        if (!user) {
          throw new Error(`No user found for email: ${email}`)
        }
        const passwordValid = await compare(password, user.password)
        if (!passwordValid) {
          throw new Error('Invalid password')
        }
        return {
          token: sign({ userId: user.id }, APP_SECRET),
          user,
        }
      },
    })
```
Example - try/catch:

```typescript
    t.field('login', {
      type: 'AuthPayload',
      args: {
        email: stringArg(),
        password: stringArg(),
      },
      resolve: async (parent, { email, password }, context) => {
        try { // Workaround for photon #89
          const user = await context.photon.users.findOne({
            where: {
              email,
            },
          })
          const passwordValid = await compare(password, user.password)
          if (!passwordValid) {
            throw new Error('Invalid password')
          }
          return {
            token: sign({ userId: user.id }, APP_SECRET),
            user,
          }
        } catch(e) {
          throw new Error(`Invalid user or password for email: ${email}\n${e}`)
        }
      },
    })
```
</p></details>

**Comment by ryan-cat** [comment]
+1 to have findOne return null in the case that it does not find a record. Sometimes null is acceptable and allowed. Allows us to decide whether something is an error or not and removes the need for unnecessary amounts of try-catch blocks to do so.

**Comment by schickling** [comment]
Thanks a lot for all of your great feedback. We've decided to change the default behavior of `find` to return a nullable result. This change (including others) is currently being specced out in a new [Photon API design](https://github.com/prisma/specs/blob/photonjs-improvements/specs/PhotonJS.md). Please check it out and provide more feedback for it [here](https://github.com/prisma/specs/pull/5). 🙏

**Comment by jferrettiboke** [comment]
@schickling I have to write boilerplate because of this. Please, ship it in the next release if the spec is ready. Try/catch makes code blocks look ugly, and some times it's hard to read.

I would love to see in docs how to handle errors without tears 😭.

**Comment by gihrig** [comment]
I'm finding that `try/catch` does not always catch `.findOne` errors. In those cases I'm using `findMany` when I expect only a single response.

e.g.
```ts
export async function isPost(id: string, ctx: Context) {
  // TODO: .findMany is used to avoid uncatchable error thrown
  // by findOne. see Photon #89
  const post = await ctx.photon.posts
  .findMany({
    where: {
      id
    }
  })
  if (post.length === 0) {
    return false
  }
  return true
}
```

**Comment by skyflyer55** [comment]
The workout that I use is to pass a function to the catch method of photon

```
export const NotExistError = err => (err.message.match(/RecordDoesNotExist/i) ? null : err);
```
```
resolve: (parent, { search }, ctx) => {
    return ctx.photon.users.findOne({
            where: search
    })
    .then(data => data);
    .catch(NotExistError);
}
```

Just waiting for the new change of allowing null to be returned to be pushed in one of the upcoming previews

**Comment by nhuesmann** [comment]
Hey @schickling, the first link in your latest comment is broken. Where can I find the status of this feature? I'm definitely looking forward to getting null/no results rather than an ugly error, would love to have that feature soon.

**Comment by schickling** [comment]
@nhuesmann fixed in comment above. This should be implemented fairly soon!

**Comment by nhuesmann** [comment]
@schickling thanks! I do hope it's soon, I'm currently maintaining a bit of excess code to catch all these.

Side note: in the doc you linked, the model name is changed to singular (`photon.users` -> `photon.user`) and the single record query is changed to `find` from `findOne`. Are these the anticipated final forms?

**Comment by owcc** [comment]
hope this fixed in the next release! I hit this when I do query from a table without record, I use `findMany` instead of `findOne` temporarily to avoid throw `photon error`.

**Comment by AbdusamadTurdiev** [comment]
@nhuesmann @schickling hope that `findOne` will return null if there's no record because I'm using this poor workaround for now

![image](https://user-images.githubusercontent.com/33511684/65965767-934cc080-e478-11e9-8d50-150a2f08a9ff.png)

**Comment by cypcz** [comment]
When is this going to be introduced?

**Comment by divyenduz** [comment]
Closing in favor of https://github.com/prisma/photonjs/issues/245
